File: web/test/appearance.test.mjs
// Copyright (c) 2026 anansi
//
// Part of a fork of web-terminal (https://github.com/code-yeongyu/web-terminal),
// Copyright (c) 2026 YeonGyu Kim, MIT licensed.
// Full license text: LICENSE at the repo root.

import { strict as assert } from "node:assert";
import {
  DEFAULT_ALPHA,
  DEFAULT_FONT,
  DEFAULT_FONT_SIZE,
  DEFAULT_THEME,
  DELTA_KEY,
  FONTS,
  THEMES,
  clampSize,
  fontStack,
  nextSize,
  readDelta,
  resolveAppearance,
  sizeHotkey,
  writeDelta,
} from "../appearance.js";

const mem = () => {
  const m = new Map();
  return {
    getItem: (k) => (m.has(k) ? m.get(k) : null),
    setItem: (k, v) => { m.set(k, String(v)); },
    removeItem: (k) => { m.delete(k); },
  };
};

assert.equal(resolveAppearance({}).theme, DEFAULT_THEME);
assert.equal(resolveAppearance({}).font, DEFAULT_FONT);
assert.equal(resolveAppearance({}).fontSize, DEFAULT_FONT_SIZE);
assert.equal(resolveAppearance({}).alpha, DEFAULT_ALPHA);
assert.equal(resolveAppearance({ theme: "solarized", font: "comic", fontSize: "x" }).theme, DEFAULT_THEME);
assert.equal(resolveAppearance({ theme: "gruvbox-dark", font: "iosevka", fontSize: 16 }).font, "iosevka");
assert.equal(resolveAppearance({ theme: "gruvbox-dark", font: "iosevka", fontSize: 16 }).fontSize, 16);
assert.equal(resolveAppearance({ alpha: true }).alpha, true);
assert.equal(resolveAppearance({ alpha: "true" }).alpha, false);
assert.equal(resolveAppearance({ alpha: 1 }).alpha, false);

assert.equal(clampSize(14, 0), 14);
assert.equal(clampSize(14, 20), 22);
assert.equal(clampSize(14, -20), 10);

{
  const s = mem();
  assert.equal(readDelta(s), 0);
  assert.equal(nextSize("inc", 14, s), 15);
  assert.equal(s.getItem(DELTA_KEY), "1");
  assert.equal(nextSize("dec", 14, s), 14);
  assert.equal(s.getItem(DELTA_KEY), null);
  writeDelta(8, s);
  assert.equal(nextSize("inc", 14, s), 22);
  assert.equal(s.getItem(DELTA_KEY), "8");
  assert.equal(nextSize("reset", 14, s), 14);
  assert.equal(s.getItem(DELTA_KEY), null);
}

const key = (partial) => ({ type: "keydown", altKey: false, ctrlKey: false, metaKey: false, key: "", ...partial });
assert.equal(sizeHotkey(key({ ctrlKey: true, key: "=" })), "inc");
assert.equal(sizeHotkey(key({ metaKey: true, key: "+" })), "inc");
assert.equal(sizeHotkey(key({ ctrlKey: true, key: "-" })), "dec");
assert.equal(sizeHotkey(key({ ctrlKey: true, key: "0" })), "reset");
assert.equal(sizeHotkey(key({ key: "-" }), { ctrlLatched: true }), "dec");
assert.equal(sizeHotkey(key({ key: "=" }), { ctrlLatched: true }), "inc");
assert.equal(sizeHotkey(key({ ctrlKey: true, altKey: true, key: "=" })), null);
assert.equal(sizeHotkey(key({ key: "c" }), { ctrlLatched: true }), null);
assert.equal(sizeHotkey(key({ key: "-" })), null);

assert.ok(THEMES[DEFAULT_THEME].background);
assert.ok(FONTS[DEFAULT_FONT].file.endsWith(".woff2"));
assert.equal(fontStack("Lilex Nerd Font Mono"), `"Lilex Nerd Font Mono", ui-monospace, monospace`);

for (const theme of Object.values(THEMES)) {
  for (const k of [
    "background", "foreground", "cursor", "cursorAccent",
    "black", "red", "green", "yellow", "blue", "magenta", "cyan", "white",
    "brightBlack", "brightRed", "brightGreen", "brightYellow",
    "brightBlue", "brightMagenta", "brightCyan", "brightWhite",
  ]) {
    assert.equal(typeof theme[k], "string", k);
    assert.match(theme[k], /^#[0-9a-fA-F]{6}$/, k);
  }
}

console.log("ok");

File: internal/web/appearance_test.go
// Copyright (c) 2026 anansi
//
// Part of a fork of web-terminal (https://github.com/code-yeongyu/web-terminal),
// Copyright (c) 2026 YeonGyu Kim, MIT licensed. The Go backend, protocol design
// and frontend architecture in this fork derive from that project; this file is
// new work. Full license text: LICENSE at the repo root.

package web

import "testing"

func TestLoadAppearanceDefaults(t *testing.T) {
	t.Setenv("WT_THEME", "")
	t.Setenv("WT_FONT", "")
	t.Setenv("WT_FONT_SIZE", "")
	t.Setenv("WT_ALPHA", "")
	got := LoadAppearance()
	if got.Theme != defaultTheme || got.Font != defaultFont || got.FontSize != defaultFontSize || got.Alpha {
		t.Fatalf("defaults = %+v", got)
	}
}

func TestLoadAppearanceValid(t *testing.T) {
	t.Setenv("WT_THEME", "gruvbox-dark")
	t.Setenv("WT_FONT", "iosevka")
	t.Setenv("WT_FONT_SIZE", "16")
	t.Setenv("WT_ALPHA", "1")
	got := LoadAppearance()
	if got.Theme != "gruvbox-dark" || got.Font != "iosevka" || got.FontSize != 16 || !got.Alpha {
		t.Fatalf("got %+v", got)
	}
}

func TestLoadAppearanceAlpha(t *testing.T) {
	t.Setenv("WT_ALPHA", "true")
	if !LoadAppearance().Alpha {
		t.Fatal("true")
	}
	t.Setenv("WT_ALPHA", "off")
	if LoadAppearance().Alpha {
		t.Fatal("off")
	}
}

func TestLoadAppearanceUnknownFallsBack(t *testing.T) {
	t.Setenv("WT_THEME", "solarized")
	t.Setenv("WT_FONT", "comic-sans")
	t.Setenv("WT_FONT_SIZE", "nope")
	got := LoadAppearance()
	if got.Theme != defaultTheme || got.Font != defaultFont || got.FontSize != defaultFontSize {
		t.Fatalf("fallback = %+v", got)
	}
}
